### PR TITLE
Use dependency injection to access the current ruleset

### LINKED ISSUE #2847: Changing mode after playing osu!mania with nonstandard number of keys crashes game
Played osu!mania with different number of keys, and changed mode to osu!. Game freezed for few seconds and crashed.

It does the same with other game modes.

logs:
[performance-gamethread.draw.log](https://github.com/ppy/osu/files/2121212/performance-gamethread.draw.log)
[performance-gamethread.input.log](https://github.com/ppy/osu/files/2121213/performance-gamethread.input.log)
[performance-gamethread.update.log](https://github.com/ppy/osu/files/2121214/performance-gamethread.update.log)
[runtime.log](https://github.com/ppy/osu/files/2121215/runtime.log)
[database.log](https://github.com/ppy/osu/files/2121216/database.log)
[network.log](https://github.com/ppy/osu/files/2121217/network.log)
[performance.log](https://github.com/ppy/osu/files/2121218/performance.log)
[performance-gamethread.audio.log](https://github.com/ppy/osu/files/2121219/performance-gamethread.audio.log)

Playing on win 10 if it changes anything.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/59929217-changing-mode-after-playing-osu-mania-with-nonstandard-number-of-keys-crashes-game?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by peppy**
Previously this was done directly from `OsuGame`, which is something we're moving away from. It also didn't offer the flexibility to fix a bug where the leaderboard would attempt to retrieve scores from an invalid beatmap/ruleset combination.

- Closes #2847.

---

Fixes error notifications / hard crashes when switching the active ruleset at song select.

### CODE REVIEW

File: osu.Game/Overlays/Toolbar/ToolbarModeSelector.cs [hunk 1]
```diff
@@ -83,8 +83,8 @@ private void load(RulesetStore rulesets, OsuGame game)
             ruleset.ValueChanged += rulesetChanged;
             ruleset.DisabledChanged += disabledChanged;
 
-            if (game != null)
-                ruleset.BindTo(game.Ruleset);
+            if (ruleset != null)
```
Comment (smoogipoo) [inline osu.Game/Overlays/Toolbar/ToolbarModeSelector.cs]: Incorrect conditional

File: osu.Game/Screens/Select/SongSelect.cs [hunk 2]
```diff
@@ -62,6 +64,8 @@ public abstract class SongSelect : OsuScreen
         private SampleChannel sampleChangeDifficulty;
         private SampleChannel sampleChangeBeatmap;
 
+        protected new readonly Bindable<RulesetInfo> Ruleset = new Bindable<RulesetInfo>();
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/SongSelect.cs]: Are you sure we don't want to be re-DI'ing this ruleset in and replacing the existing one? Seems like components underneath songselect (leaderboard, wedge) should use this one, no?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Select/SongSelect.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Select/SongSelect.cs]: ah yeah, that got lost in some refactoring. that was the whole point of this PR

File: osu.Game/OsuGame.cs [hunk 3]
```diff
@@ -85,7 +85,7 @@ private Intro intro
         private OnScreenDisplay onscreenDisplay;
 
         private Bindable<int> configRuleset;
-        public Bindable<RulesetInfo> Ruleset = new Bindable<RulesetInfo>();
+        private readonly Bindable<RulesetInfo> ruleset = new Bindable<RulesetInfo>();
```
Comment (smoogipoo) [inline osu.Game/OsuGame.cs]: For now, I'd really like if we also exposed an `IBindable<RulesetInfo>` as a dependency, and then use that one wherever the ruleset isn't actually modified.

This way I think it'll ease our pain when we later remove the `Bindable<Ruleset>` from dependencies.

File: osu.Game/Screens/Select/SongSelect.cs [hunk 4]
```diff
@@ -280,6 +283,7 @@ void performLoad()
                     bool preview = beatmap?.BeatmapSetInfoID != Beatmap.Value?.BeatmapInfo.BeatmapSetInfoID;
 
                     Beatmap.Value = beatmaps.GetWorkingBeatmap(beatmap, Beatmap.Value);
+                    Ruleset.Value = base.Ruleset.Value;
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/SongSelect.cs]: This doesn't get invoked if the ruleset is changed via the toolbar.

File: osu.Game/Screens/Select/SongSelect.cs [hunk 5]
```diff
@@ -176,10 +180,12 @@ protected SongSelect()
             }
         }
 
+
```
Comment (smoogipoo) [inline osu.Game/Screens/Select/SongSelect.cs]: Trim

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Should be good now.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**